### PR TITLE
Return an array of error for request validation

### PR DESCRIPTION
**Pull request by blacksponge**
## Purpose

This PR changes how error related to param validation are return to the client to allow better analysis client side. It also prevent the internal error due to failed validation.

## Context

Replaces #795.

Addresses #703, is a requirement for https://github.com/zonemaster/zonemaster-gui/issues/177 and https://github.com/zonemaster/zonemaster-gui/issues/178. Dependency for  #230.

## Changes

* The json validation error are split in an array of object to allow better parsing client side (thus enabling better error message on the GUI). The json validation error are mapped to a set of custom defined messages to allow more user friendly messages but also translation.
 
## How to test this PR

`curl -sH "Content-Type: application/json" --data '{"method":"start_domain_test","params":{"nameservers":[{"ns":"/!\\", "ip":":::1"}], "ds_info":[{"algorithm":-1, "digest":"not a keytag"}], "profile":"toto", "client_id":"toto", "domain":"dfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdiis.se"},"id":1,"jsonrpc":"2.0"}' localhost:5000/api  | jq`
```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "error": {
    "message": "Invalid method parameter(s).",
    "code": "-32602",
    "data": [
      {
        "path": "/ds_info/0/algorithm",
        "message": "-1 < minimum(0)"
      },
      {
        "message": "String does not match (?^ui:^[a-f0-9]{40}$|^[a-f0-9]{64}$|^[a-f0-9]{96}$).",
        "path": "/ds_info/0/digest"
      },
      {
        "path": "/ds_info/0/digtype",
        "message": "Missing property."
      },
      {
        "message": "Missing property.",
        "path": "/ds_info/0/keytag"
      },
      {
        "path": "/profile",
        "message": "Unknown profile"
      },
      {
        "message": "The domain name or label is too long",
        "path": "/domain"
      },
      {
        "message": "The domain name character(s) are not supported",
        "path": "/nameservers/0/ns"
      },
      {
        "message": "Invalid IP address",
        "path": "/nameservers/0/ip"
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by matsduf** [comment]
This PR has many commits, including several merge commits. Is it possible reduce the number of commits here?

**Comment by blacksponge** [comment]
> Is it possible reduce the number of commits here?

We could always squash them when will merge

**Comment by matsduf** [comment]
In a large PR the individual commits can help your orientation, but these are too many for that. I am thinking of review time rather than merge time. And also, there are six merge commits in this PR.

**Comment by blacksponge** [comment]
> Is it possible reduce the number of commits here?

all commit have been squashed.

File: lib/Zonemaster/Backend/RPCAPI.pm [hunk 1]
```diff
@@ -29,6 +29,7 @@ use Zonemaster::Backend::Validator;
 
 my $zm_validator = Zonemaster::Backend::Validator->new;
 my %json_schemas;
+our %extra_validators;
```
Comment (matsduf) [inline lib/Zonemaster/Backend/RPCAPI.pm:32]: Any reason for "our" instead of "my"?

Comment (blacksponge) [reply lib/Zonemaster/Backend/RPCAPI.pm:32]: I need to access this from the tests, I am not sure if an accessor function would be better or if that is okay to do?

Comment (matsduf) [reply lib/Zonemaster/Backend/RPCAPI.pm:32]: It is only used in RPCAPI.pm, isn't it? And that is the same lexical scope.

Comment (blacksponge) [reply lib/Zonemaster/Backend/RPCAPI.pm:32]: It is used RPCAPI.pm and test_validate_syntax.t https://github.com/zonemaster/zonemaster-backend/blob/dd09207b2d9dccd5a0a4ded68890e56edc6f4937/t/test_validate_syntax.t#L33

File: docs/API.md [hunk 2]
```diff
@@ -694,12 +736,50 @@ The parameters that are compared when to determine if two requests are to be
...
+        "message": "Expected integer - got string.",
+        "path": "/ds_info/0/algorithm"
+      },
+      {
+        "message": "Missing property.",
+        "path": "/ds_info/0/digest"
+      },
+      {
+        "path": "/profile",
+        "message": "Unknown profile"
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply docs/API.md:775]: I have update the profile section to contains the new error messages.

File: docs/API.md [hunk 3]
```diff
@@ -139,22 +144,19 @@ Basic data type: object
...
 
 Basic data type: string
 
-This parameter is a string that are an IPv4 or IPv6. It's validated with the following regexes:
- - IPv4 : `/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/`
- - IPv6 : `/^([0-9A-Fa-f]{1,4}:[0-9A-Fa-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)|([0-9A-Fa-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/`
-
+This parameter is a string that is either
+ - a valid IPv4 in [dot-decimal notation] ;
+ - a valid IPv6 in [recommend text format for IPv6 addresses].
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply docs/API.md]: Corrected

File: lib/Zonemaster/Backend/RPCAPI.pm (hunk 1 above)
Comment (matsduf) [reply lib/Zonemaster/Backend/RPCAPI.pm:32]: Ok.

**Review by matsduf — APPROVED** [review]

File: t/test01.t [hunk 4]
```diff
@@ -135,12 +135,6 @@ sub run_zonemaster_test_with_backend_API {
     ok( defined $test_results->{creation_time},      'TEST1 $test_results->{creation_time} defined' );
     ok( defined $test_results->{results},            'TEST1 $test_results->{results} defined' );
     cmp_ok( scalar( @{ $test_results->{results} } ), '>', 1, 'TEST1 got some results' );
-
-    dies_ok { $engine->get_test_results( { id => $hash_id, language => 'fr-FR' } ); }
-    'API get_test_results -> [results] parameter not present (wrong language tag: underscore not hyphen)'; # Should be underscore, not hyphen.
-
-    dies_ok { $engine->get_test_results( { id => $hash_id, language => 'zz' } ); }
-    'API get_test_results -> [results] parameter not present (wrong language tag: "zz" unknown)'; # "zz" is not our configuration file.
```
Comment (pnax) [inline t/test01.t:143]: Why did you remove these tests?

File: t/test_validate_syntax.t [hunk 5]
```diff
@@ -221,21 +222,24 @@ subtest 'Everything but NoWarnings' => sub {
 
     $frontend_params->{ds_info}->[0]->{algorithm} = 1;
     $frontend_params->{ds_info}->[0]->{digest}    = '0123456789012345678901234567890123456789';
-    is( $engine->validate_syntax( $frontend_params )->{status}, 'ok', encode_utf8( 'Valid Algorithm Type [numeric format]' ) )
-        or diag( $engine->validate_syntax( $frontend_params )->{message} );
+    $frontend_params->{ds_info}->[0]->{digtype}   = 1;
+    $frontend_params->{ds_info}->[0]->{keytag}   = 5000;
```
Comment (pnax) [inline t/test_validate_syntax.t:226]: Would you mind adding tests with `digtype` and `keytag`?

File: t/test01.t (hunk 4 above)
Comment (blacksponge) [reply t/test01.t:143]: That change should be moved to #823.

File: t/test_validate_syntax.t (hunk 5 above)
Comment (blacksponge) [reply t/test_validate_syntax.t:226]: done.

File: docs/API.md [hunk 6]
```diff
@@ -591,10 +629,9 @@ Example response:
 
 #### `"params"`
 
-An object with the property:
-
-`"domain"`: A *domain name*, required. The domain whose DNS records are requested.
+An object with the properties:
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply docs/API.md]: fixed

File: docs/API.md [hunk 7]
```diff
@@ -507,9 +529,9 @@ Example response:
 
 #### `"params"`
 
-An object with the property:
+An object with the properties:
```
[the remark this answers is not in the document]
Comment (blacksponge) [reply docs/API.md]: fixed

**Review by pnax — APPROVED** [review]

**Comment by pnax** [comment]
Would you say that your PR resolves #637?

**Comment by blacksponge** [comment]
> Would you say that your PR resolves #637?

Most of it, I think, yes.

**Review by matsduf — APPROVED** [review]

**Comment by mattias-p** [comment]
I've tested this for v2021.2 and it LGTM.
